qubes/ext/utils.py

```diff
@@ -100,3 +138,19 @@ def compare_device_cache(vm, devices_cache, current_devices):
             if cached_front is not None:
                 detached[dev_id] = cached_front
     return added, attached, detached, removed
+
+
+def confirm_device_attachment(device, frontends) -> str:
+    try:
+        # pylint: disable=consider-using-with
+        proc = subprocess.Popen(
+            ["attach-confirm", device.backend_domain.name,
+            device.port_id, device.description,
+             *[f.name for f in frontends.keys()]],
+            stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+        (target_name, _) = proc.communicate()
+        return target_name.decode()
+    except Exception as exc:
+        print("attach-confirm", exc, file=sys.stderr)
+        return ""
```

Comment:
There are a few problems with this function:
1. It must be async, since it may block for some time, and it's not acceptable to block the whole qubesd for this time. In fact, attach-confirm probably won't work this way at all if it tries to talk to qubesd, since it's blocked on waiting for attach-confirm...
2. subprocess.Popen -> asyncio variant
3. The tool name is IMO too generic for a tool in a common /usr/bin/
4. The tool belongs to desktop-linux-manager repo, which looks like a layering violation - dom0 code should also work without any of the GUI frontends installed.
5. Extension of the above: ensure it will work also with GUI domain. See how policy prompts are done for example. And this also will need adjustment to the threat model: verify the response is one of allowed ones (on the `frontends` list?)

Reply:
I see the code of `attach-confirm` already is async. Maybe simply put its code here, instead of calling external program? It means you will need to keep the `params` dict format in sync, but changes there needs to be done in compatible way anyway (due to the GUI domain case, where both ends may be updated independently). Plus, you won't need to make external `get_system_info()` call, as by running inside qubesd you already have all the info here.

Reply:
> * The tool belongs to desktop-linux-manager repo, which looks like a layering violation - dom0 code should also work without any of the GUI frontends installed in dom0.

I agree that this to some level introduces a dependency of the lower layer on the upper one, and that's why I decided to implement it as an independent program. If we can think of another way to confirm device attaching, it should be easy to implement. Currently, however, the GUI is required to implement `ask-to-attach`, and if the package is not available, such an assignment will simply be ignored. So the implementation is now: for `ask-to-attach`, ask the user for confirmation if possible, otherwise ignore the assignment.

Reply:
For sys-gui case, since this is socket protocol already, it's very easy to pipe it to a qrexec service. As said above, I'd prefer the interface between those two packages be "send this to a socket, read response in this format", instead of "call this executable". It will allow adding qrexec support without needing to still have the GUI-related package installed in dom0.

And yes, for now the approach "deny if confirmation not available" is fine.